source/s3_meta_request.c

```diff
@@ -825,62 +829,105 @@ void aws_s3_meta_request_stream_response_body_synced(
     uint32_t num_streaming_requests = 0;
 
     /* Grab any additional requests that could be streamed to the caller. */
     while (next_streaming_request != NULL) {
-        aws_atomic_fetch_sub(&client->stats.num_requests_queued_waiting, 1);
+        aws_atomic_fetch_sub(&client->stats.num_requests_stream_queued_waiting, 1);
 
         aws_linked_list_push_back(&streaming_requests, &next_streaming_request->node);
         ++num_streaming_requests;
-        next_streaming_request = aws_s3_meta_request_body_streaming_pop_synced(meta_request);
+        next_streaming_request = s_s3_meta_request_body_streaming_pop_next_synced(meta_request);
     }
 
-    if (!aws_linked_list_empty(&streaming_requests)) {
-        aws_s3_client_stream_response_body(
-            client, meta_request, &streaming_requests, s_s3_meta_request_streaming_body_callback, meta_request);
-
-        meta_request->synced_data.num_parts_delivery_sent += num_streaming_requests;
+    if (aws_linked_list_empty(&streaming_requests)) {
+        return;
     }
+
+    aws_atomic_fetch_add(&client->stats.num_requests_streaming, num_streaming_requests);
+
+    meta_request->synced_data.num_parts_delivery_sent += num_streaming_requests;
+
+    struct s3_stream_response_body_payload *payload =
+        aws_mem_calloc(client->sba_allocator, 1, sizeof(struct s3_stream_response_body_payload));
+
+    aws_s3_meta_request_acquire(meta_request);
+    payload->meta_request = meta_request;
+
+    aws_linked_list_init(&payload->requests);
+    aws_linked_list_swap_contents(&payload->requests, &streaming_requests);
+
+    aws_task_init(
+        &payload->task, s_s3_meta_request_body_streaming_task, payload, "s_s3_meta_request_body_streaming_task");
+    aws_event_loop_schedule_task_now(meta_request->io_event_loop, &payload->task);
 }
 
-static void s_s3_meta_request_delivered_requests(
-    struct aws_s3_meta_request *meta_request,
-    int error_code,
-    uint32_t num_failed,
-    uint32_t num_successful) {
+static void s_s3_meta_request_body_streaming_task(struct aws_task *task, void *arg, enum aws_task_status task_status) {
+    (void)task;
+    (void)task_status;
+
+    struct s3_stream_response_body_payload *payload = arg;
+    AWS_PRECONDITION(payload);
+
+    struct aws_s3_meta_request *meta_request = payload->meta_request;
     AWS_PRECONDITION(meta_request);
     AWS_PRECONDITION(meta_request->vtable);
-    AWS_PRECONDITION(meta_request->vtable->delivered_requests);
 
-    meta_request->vtable->delivered_requests(meta_request, error_code, num_failed, num_successful);
-}
+    struct aws_s3_client *client = meta_request->client;
+    AWS_PRECONDITION(client);
 
-void aws_s3_meta_request_delivered_requests_default(
-    struct aws_s3_meta_request *meta_request,
-    int error_code,
-    uint32_t num_failed,
-    uint32_t num_successful) {
+    /* Client owns this event loop group. A cancel should not be possible. */
```

Comment:
Are we sure about this? Does that imply that the client would wait for all tasks to complete somewhere before shutting down?

Reply:
Yes--the meta request holds a reference to the client while the meta request is in progress, so the related ELG will never get shutdown until the meta request is finished, which indirectly causes the "wait".

If the meta request gets canceled and a task has been scheduled, the task still needs to run before the meta request can finish (and the client to clean up).  But, it'll just throw away the request in that case instead of trying to stream it to the user.